File: grpc/cpp/receiver_async_test.cpp
#include <iostream>
#include <memory>
#include <string>
#include <signal.h>
#include <grpcpp/grpcpp.h>
#include "test_data.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
using test_data::TestDataItem;
using test_data::Ack;
using test_data::TestDataService;

class TestDataServiceImpl final : public TestDataService::Service {
public:
    TestDataServiceImpl(int receiver_id) : receiver_id_(receiver_id), messages_received_(0) {}

    Status TransferData(ServerContext* context, const TestDataItem* request,
                        Ack* reply) override {
        messages_received_++;
        std::cout << " [Receiver " << receiver_id_ << "] [ASYNC] Received message " << request->message_id() << std::endl;
        reply->set_status("ACK");
        reply->set_message_id(request->message_id());
        reply->set_receiver_id(receiver_id_);
        return Status::OK;
    }

    int get_messages_received() const { return messages_received_; }

private:
    int receiver_id_;
    int messages_received_;
};

std::unique_ptr<Server> server_ptr;

void signal_handler(int sig) {
    if (server_ptr) {
        server_ptr->Shutdown();
    }
}

void RunServer(int receiver_id) {
    int port = 50051 + receiver_id;
    std::string server_address("0.0.0.0:" + std::to_string(port));
    TestDataServiceImpl service(receiver_id);

    ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    server_ptr = builder.BuildAndStart();
    std::cout << " [*] [ASYNC] Receiver " << receiver_id << " listening on " << server_address << std::endl;
    server_ptr->Wait();
    std::cout << " [x] [ASYNC] Receiver " << receiver_id << " shutting down (received " << service.get_messages_received() << " messages)" << std::endl;
}

int main(int argc, char** argv) {
    int receiver_id = 0;
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "--id" && i + 1 < argc) {
            receiver_id = std::stoi(argv[i + 1]);
        }
    }

    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    RunServer(receiver_id);
    return 0;
}

File: grpc/python/receiver_async_test.py
#!/usr/bin/env python3
"""gRPC Python Receiver - Async"""
import sys
import signal
import asyncio
import grpc
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

import messaging_pb2
import messaging_pb2_grpc
from message_helpers import *

class AsyncMessagingServicer(messaging_pb2_grpc.MessagingServiceServicer):
    def __init__(self, receiver_id):
        self.receiver_id = receiver_id
        
    async def SendMessage(self, request, context):
        message_id = request.message_id
        print(f" [x] [ASYNC] Received message {message_id}")
        
        # Create ACK
        response = create_ack_from_envelope(request, str(self.receiver_id))
        setattr(response, 'async', True)
        return response

async def run(receiver_id, server_port):
    server = grpc.aio.server()
    messaging_pb2_grpc.add_MessagingServiceServicer_to_server(
        AsyncMessagingServicer(receiver_id), server
    )
    port = server_port + receiver_id
    server.add_insecure_port(f'[::]:{port}')
    
    print(f" [*] [ASYNC] Receiver {receiver_id} listening on port {port}")
    await server.start()
    
    try:
        await server.wait_for_termination()
    except asyncio.CancelledError:
        await server.stop(0)

def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--id', type=int, default=0)
    parser.add_argument('--server-port', type=int, default=50051)
    args = parser.parse_args()
    
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    
    task = loop.create_task(run(args.id, args.server_port))
    
    def signal_handler():
        task.cancel()
        
    loop.add_signal_handler(signal.SIGINT, signal_handler)
    loop.add_signal_handler(signal.SIGTERM, signal_handler)
    
    try:
        loop.run_until_complete(task)
    except (KeyboardInterrupt, asyncio.CancelledError):
        pass
    finally:
        loop.close()


if __name__ == "__main__":
    main()
